pkg/cli/loader/cluster.go

```diff
@@ -0,0 +1,291 @@
+		return nil, fmt.Errorf("loader is closed")
+	}
+	cl.mutex.RUnlock()
+
+	options = cl.applyDefaults(options)
+
+	if err := cl.validateOptions(options); err != nil {
+		return nil, fmt.Errorf("invalid options: %w", err)
+	}
+
+	tasks := cl.createLoadingTasks(options.ResourceTypes, options)
+
+	results, err := cl.executeTasks(ctx, tasks, options)
+	if err != nil {
+		return nil, fmt.Errorf("task execution failed: %w", err)
+	}
+
+	finalResult := cl.aggregateResults(results, startTime, options.Concurrency)
+
+	return finalResult, nil
+}
+
+func (cl *ClusterLoader) applyDefaults(options ResourceOptions) ResourceOptions {
+	if options.Concurrency <= 0 {
+		options.Concurrency = cl.config.DefaultConcurrency
+	}
+	if options.BatchSize <= 0 {
+		options.BatchSize = cl.config.DefaultBatchSize
+	}
+	if options.Timeout <= 0 {
+		options.Timeout = cl.config.DefaultTimeout
+	}
+	if options.MaxRetries <= 0 {
+		options.MaxRetries = cl.config.DefaultMaxRetries
+	}
+	return options
+}
+
+func (cl *ClusterLoader) validateOptions(options ResourceOptions) error {
+	if len(options.ResourceTypes) == 0 {
+		return fmt.Errorf("at least one resource type must be specified")
+	}
+
+	if options.Concurrency > 32 {
+		return fmt.Errorf("concurrency cannot exceed 32")
+	}
+
+	if options.BatchSize > 2000 {
+		return fmt.Errorf("batch size cannot exceed 2000")
+	}
+
+	return nil
+}
+
+func (cl *ClusterLoader) createLoadingTasks(gvks []schema.GroupVersionKind, options ResourceOptions) []LoadTask {
+	var tasks []LoadTask
+	taskID := 0
+
+	for _, gvk := range gvks {
+		gvr := schema.GroupVersionResource{
+			Group:    gvk.Group,
+			Version:  gvk.Version,
+			Resource: cl.pluralizeKind(gvk.Kind),
+		}
+
+		client := cl.client.Resource(gvr)
+
+		listOptions := metav1.ListOptions{
+			Limit: int64(options.BatchSize),
+		}
+
+		if options.LabelSelector != "" {
+			listOptions.LabelSelector = options.LabelSelector
+		}
+
+		if options.FieldSelector != "" {
+			listOptions.FieldSelector = options.FieldSelector
+		}
+
+		if options.Namespace != "" {
+			tasks = append(tasks, LoadTask{
+				ID:          fmt.Sprintf("task-%d", taskID),
+				GVK:         gvk,
+				GVR:         gvr,
+				Namespace:   options.Namespace,
+				ListOptions: listOptions,
+				Client:      client.Namespace(options.Namespace),
+			})
+		} else {
+			tasks = append(tasks, LoadTask{
+				ID:          fmt.Sprintf("task-%d", taskID),
+				GVK:         gvk,
+				GVR:         gvr,
+				Namespace:   "",
+				ListOptions: listOptions,
+				Client:      client,
+			})
+
+		}
+		taskID++
+	}
+
+	return tasks
+}
+
+func (cl *ClusterLoader) executeTasks(ctx context.Context, tasks []LoadTask, options ResourceOptions) ([]LoadTaskResult, error) {
+	if len(tasks) == 0 {
+		return nil, nil
+	}
+
+	if options.Concurrency != cl.workerPool.workers {
+		cl.workerPool.Close()
+		cl.workerPool = NewWorkerPool(WorkerPoolConfig{
+			Workers:   options.Concurrency,
+			QueueSize: options.Concurrency * 2,
+			Logger:    cl.logger,
+		})
```

Comment:
I noticed that we already initialize a worker pool when creating a new cluster loader:
https://github.com/kyverno/kyverno/blob/9f9be7b5f279a94e9a1e46fa662393fbfad62995/pkg/cli/loader/cluster.go#L59-L63

Instead of checking the concurrency value against the number of workers, closing the existing pool, and starting a new one, I suggest we initialize the worker pool just once during the creation of the cluster loader. Since the concurrency value remains constant throughout execution, there's no need to recreate the pool.